chapters/operatorsandexpressions.tex

```diff
@@ -1414,6 +1414,10 @@ \section{Variability of Expressions}\label{variability-of-expressions}
 end Test;
 \end{lstlisting}
 \end{example}
+\subsection{Function Variability}\label{function-variability}
+The variability of function calls need to consider both the variability of arguments directly given in the function and the variability of the used default arguments, if any.
```

Comment:
Good start, but I believe it is possible to be more concrete here and leave less or no gaps in the specification.  I'd also like to attempt an approach which is somewhat closer to something one could also imagine being expressed by the function interface.

To fill the gaps, I think we need a bit of terminology:
```suggestion
The variability of function calls need to consider both the variability of arguments directly given in the function and the variability of the used default arguments, if any.
The \firstuse{variability of a function} is defined as the highest variability among all default argument expressions.
```

Then, we would change a formulation like
> … a function or operator with parameter subexpressions is a parameter expression.
into
> … an operator or function call with parameter function variability and parameter subexpressions, is a parameter expression.

(It should be done similarly for constant and discrete-time expressions.)

Reply:
That doesn't seem to work with the formulation as only the variability of the _used_ default arguments matter.
I know that it is a minor point and I can't see any practical use, but I think it would be difficult to explain that since `x` in `foo` had a parameter-default then `foo(x=2)` has parameter-variability.

Reply:
To me this is actually a major point, so let's make sure we think about it carefully, ideally with the help of others – maybe even a phone meeting in the end.

For me, assigning parameter variability to `foo(x = 2)` was a price I was happy to pay for a much simpler concept of what the variability of a function is: instead of having to keep track of the variability of every single argument's default, all one has to keep track of is a single variability that bounds them all.

From such a definition that only describes what we can express with current syntax, the step is not so big if we wanted to replace the inference of a hidden function variability by an explicit declaration of function variability in the future:
```
model M
  function foo
    output Real y;
  protected
    Real z;
  algorithm
    y := z;
  end foo;
  Real x = time;
  parameter function f1 = foo(z = x); // Error: foo(z = x) has continuous-time function variability.
  parameter Real z = f1(); // OK: parameter expression.
end M;
```

Note that introducing such syntax is not on the table here.  However, it corresponds to something that I find easy to reason about, and to a logic that could be nicely supported by function variability prefixes in the future.

Reply:
What about something like this:
```Modelica
model M
  function f
    input Real x;
    input Real y = x;
    output Real z = x * y;
  end f;
    
  constant Real x = f(2.0);
end M;
```
The default arguments are allowed to depend on the input arguments, so in this case it doesn't seem possible to determine the variability of the function `f`.

Reply:
True, that's an additional issue. Clearly it doesn't influence the variability of the actual function calls (since we already consider the variability of the given input x), but it needs to be considered.

Reply:
I agree with @HansOlsson that it needs to be considered, and that it seems straight-forward how to handle it.

I am thinking of a function's variability as the variability that the function itself can contribute with in a function call.  From this it follows naturally that formal input parameters do not contribute with variability to the function itself.

Reply:
Then how is the variability of the function `f` in my example computed? Should the `x` in the default argument of `y` be ignored because it's an input argument?

Reply:
Probably, yes. 12.4.1 suggests that you first fill the slots by substituting the default values for the actual ones and then perform the type checking. But it also says the value is substituted and that may not be known until runtime.

Reply:
I am trying to avoid a solution where the slot filling process of 12.4.1 is involved in the variability definition.  By associating a single variability for the entire function (bounding the variability of all default expressions), one avoids the need to work out the slot filling process in the head when working out the variability of a function call.  (A tool might still be able to determine an effective variability of the call expression that is lower than the variability according to the specification, and may use this for optimization purposes, but shall not use it to verify variability requirements.)

> Then how is the variability of the function `f` in my example computed? Should the `x` in the default argument of `y` be ignored because it's an input argument?

Yes, but as "ignoring" a subexpression may not sound well in specification text, it is probably better to say that formal input parameters should be given constant variability when a function's variability is determined.

Reply:
Ok, though I'm not sure why you consider the slot filling problematic in this context. That's what we already implement in OpenModelica (and Dymola too it seems), and having a function variability would actually be a more complicated solution for us.

Reply:
The point of avoiding the slot filling process is:
- We can define variability based on what is visible in the actual code, which makes the concept much more transparent to users.
- This approach leads to a concept that we can later strengthen using variability prefixes, so that function variability is handled in the same direct way as other component variabilities.  Having declared function variability would both make it easier to pinpoint the location of a variability error, and add further transparency to the expression variability semantics.

I can understand your counterargument based on what is implemented in Dymola and OpenModelica, but while I am aware that I am proposing a solution that would involve some extra work on your side, I'm also highlighting that the analysis you're doing will still be useful to determine a function's effective variability, which will be necessary to take full advantage of low variability when generating code for efficient simulation.

In case someone wonders what I mean with "effective variability" as opposed to the specified variability (which I also like to think of as the "declared variability", since it should be derived from variability declarations in my mind), consider this example:
```
Real x = 1.0; /* Constant effective variability */
constant Real c = x; /* Error: Specified variability is continuous-time */
parameter Real p;
Boolean p = change(p); /* Constant effective variability */
constant Boolean cb = change(p); /* Error: Specified variability is parameter */
final parameter Real fp = 1.0; /* Constant effective variability */
```
Here, there are many opportunities for a tool to make optimizations based on constant effective variability.  Still, specified variability has to be determined in order to detect and report errors.

Reply:
>     * We can define variability based on what is visible in the actual code, which makes the concept much more transparent to users.

I'm not sure the users would agree, they'd probably be confused about why unused default arguments can cause a function to have a higher variability than expected. I'm not really opposed to the idea because of the extra work required but rather because I find it nonintuitive. It also risks breaking existing code since some user defined functions will no longer be legal in a constant context.

Reply:
> I'm not sure the users would agree, they'd probably be confused about why unused default arguments can cause a function to have a higher variability than expected. I'm not really opposed to the idea because of the extra work required but rather because I find it nonintuitive.

Yes, there are two different ideas of what is more intuitive here, so a good discussion is crucial for getting acceptance of either design.

> It also risks breaking existing code since some user defined functions will no longer be legal in a constant context.

True, but as @HansOlsson said, the risk of actually breaking real world code appears small.

Then I think I missed one important reason for avoiding the slot filling process, and that is that it corresponds so close to function evaluation that it becomes hard to handle recursive calls (please correct me if I'm wrong here), which is one of the problems to be solved in the issue we're trying to close here: https://github.com/modelica/ModelicaSpecification/issues/2538#issuecomment-604998897

This is related to https://github.com/modelica/ModelicaSpecification/pull/2924#issuecomment-851215749 below, where it is described from the perspective of just detecting variability errors, but avoids the question of how to actually determine the variability of a function.  (If function variability would be something one would declare (maybe in the future), everything would be so much easier as one would just need to check that all default argument expressions comply with the function's declared variability.)

To actually determine a function's variability without declarations is more complicated, but at least seems possible to handle using a fixed point iteration with monotonically increasing variabilities attached to the functions.  At the start of the iteration, you give all functions a lower variability bound of _constant_.  Then, as long as any resulting default argument expressions have variability exceeding it's function's lower bound, the function's lower bound is updated, and the iteration continues until convergence (convergence is no problem, and probably very rapid in reality).  After convergence, we _define_ a function's variability as the lower bound obtained from the iteration.

The complicated nature of this fixed point iteration is probably mostly an academic problem; in reality, most functions should get a variability that a human can determine easily by just looking at the function definition.  More importantly, however, is that we can get rid of this relatively complicated type-inference kind of thing that we are not used to in Modelica by introducing declared variability of functions in the future, and then we won't need to make big changes to semantics, only to syntax.

Reply:
> Then I think I missed one important reason for avoiding the slot filling process, and that is that it corresponds so close to function evaluation that it becomes hard to handle recursive calls (please correct me if I'm wrong here), which is one of the problems to be solved in the issue we're trying to close here: [#2538 (comment)](https://github.com/modelica/ModelicaSpecification/issues/2538#issuecomment-604998897)

Fair enough, though I think a simpler solution would be to just not allow such mutually recursive functions. We don't support them in OpenModelica and that's never been a problem.

Reply:
> Fair enough, though I think a simpler solution would be to just not allow such mutually recursive functions. We don't support them in OpenModelica and that's never been a problem.

(Wow, who though we'd get to hear that from Team Former MetaModelica?!)  Well, if you don't support them now, implementing the same level of support for function variability would be trivial!

Reply:
> (Wow, who though we'd get to hear that from Team Former MetaModelica?!) Well, if you don't support them now, implementing the same level of support for function variability would be trivial!

We don't support that kind of mutual recursiveness since it's not clear how it should be handled, and there's never been any user demand for it. Thus I'd prefer to forbid such constructs if it leads to a simpler specification.

Reply:
Are there other things beside the variability that aren't clear how to handle?

Reply:
Yes, the actual problem is the slot filling, you'd have to evaluate the if-expressions in that case for it to work but that's not specified (it also says that the default arguments must be acyclic in the function, though it's not clear whether that applies in this case). If we could expand such calls with the slot filling we'd be able to then calculate the variability the usual way, so that's not really an issue.

Reply:
Very interesting.  Sounds like a good topic for a separate issue that would be good to resolve before this one, don't you think?

Reply:
I have thought of the following example, but I am not sure what the current spec makes of it:
```
model M
  partial function foo
    output Real y;
  algorithm

  end foo;

  function bar = foo(y = x);
  Real x = time;
  constant Real z = foo();
end M;

```

Reply:
My intuitive understanding matches the proposed text. I would not want to have functions to have a declared variability. I think it would add unnecessary complexity to already complicated rules of Modelica. I think in cases where intuitive understanding clashes with specified rules should be avoided because they would lead to confusions and contradictions.

However, this is just my personal opinion given my current understanding of the situation. @henrikt-ma's suggestions have merit and might be beneficial in some cases. So I think it might be best to have a poll. Probably at the meeting would be the best.

One thing that I didn't see discussed is how "impure" might affect variability.